File: lib/visca.js
// VISCA-over-IP (UDP) command module for PTZOptics G3 cameras
// Default port 1259 — set VISCA_PORT in .env to override
import { createSocket } from 'dgram';

const PORT = parseInt(process.env.VISCA_PORT) || 1259;
const cl = (v, lo, hi) => Math.max(lo, Math.min(hi, Math.round(v)));

// Pan/tilt direction byte pairs: [pan_dir, tilt_dir]
// pan: 01=left, 02=right, 03=stop  |  tilt: 01=up, 02=down, 03=stop
const PDIR = {
  left:      [0x01, 0x03],
  right:     [0x02, 0x03],
  up:        [0x03, 0x01],
  down:      [0x03, 0x02],
  leftup:    [0x01, 0x01],
  rightup:   [0x02, 0x01],
  leftdown:  [0x01, 0x02],
  rightdown: [0x02, 0x02],
};

// Fire-and-forget UDP send (low latency; we don't wait for ACK)
function send(ip, buf) {
  return new Promise((resolve, reject) => {
    const sock = createSocket('udp4');
    sock.send(buf, PORT, ip, (err) => {
      sock.close();
      if (err) return reject(err);
      resolve({ url: `visca://${ip}:${PORT}`, cmd: buf.toString('hex'), ok: true, status: 200 });
    });
  });
}

export const visca = {
  // Pan/tilt: 81 01 06 01 [pan_speed] [tilt_speed] [pan_dir] [tilt_dir] FF
  pan_tilt(ip, dir, ps, ts) {
    const [pd, td] = PDIR[dir] || [0x03, 0x03];
    return send(ip, Buffer.from([0x81, 0x01, 0x06, 0x01, cl(ps,1,24), cl(ts,1,20), pd, td, 0xFF]));
  },

  // Pan/tilt stop: 81 01 06 01 01 01 03 03 FF
  stop(ip) {
    return send(ip, Buffer.from([0x81, 0x01, 0x06, 0x01, 0x01, 0x01, 0x03, 0x03, 0xFF]));
  },

  // Zoom: tele=2p, wide=3p, stop=00 (p = speed 0-7)
  zoom(ip, dir, speed) {
    const s = cl(speed, 1, 7);
    const byte = dir === 'in' ? (0x20 | s) : dir === 'out' ? (0x30 | s) : 0x00;
    return send(ip, Buffer.from([0x81, 0x01, 0x04, 0x07, byte, 0xFF]));
  },

  zoom_stop(ip) {
    return send(ip, Buffer.from([0x81, 0x01, 0x04, 0x07, 0x00, 0xFF]));
  },

  // Focus: far=2p, near=3p, stop=00
  focus(ip, dir, speed) {
    const s = cl(speed, 1, 7);
    const byte = dir === 'far' ? (0x20 | s) : dir === 'near' ? (0x30 | s) : 0x00;
    return send(ip, Buffer.from([0x81, 0x01, 0x04, 0x08, byte, 0xFF]));
  },

  focus_stop(ip) {
    return send(ip, Buffer.from([0x81, 0x01, 0x04, 0x08, 0x00, 0xFF]));
  },

  // Auto focus: 81 01 04 38 02 FF | Manual: 81 01 04 38 03 FF
  focus_mode(ip, mode) {
    return send(ip, Buffer.from([0x81, 0x01, 0x04, 0x38, mode === 'auto' ? 0x02 : 0x03, 0xFF]));
  },

  // Home: 81 01 06 04 FF
  home(ip) {
    return send(ip, Buffer.from([0x81, 0x01, 0x06, 0x04, 0xFF]));
  },

  // Preset recall: 81 01 04 3F 02 [nn] FF
  preset_recall(ip, n) {
    return send(ip, Buffer.from([0x81, 0x01, 0x04, 0x3F, 0x02, cl(n,0,254), 0xFF]));
  },

  // Preset save: 81 01 04 3F 01 [nn] FF
  preset_save(ip, n) {
    return send(ip, Buffer.from([0x81, 0x01, 0x04, 0x3F, 0x01, cl(n,0,254), 0xFF]));
  },
};

File: rackmon/visca.py
"""VISCA-over-IP (UDP) command builder + sender for PTZOptics-style cameras.

Byte layouts ported from the proven ptz-ai-controller Node implementation
(lib/visca.js): default port 1259, fire-and-forget UDP.

Builders are pure functions (unit-tested); only ViscaSender touches the
network. UDP sendto never blocks, so this is safe on the event loop.
"""

from __future__ import annotations

import logging
import socket

log = logging.getLogger("rackmon.visca")

DEFAULT_PORT = 1259

PAN_LEFT, PAN_RIGHT, PAN_STOP = 0x01, 0x02, 0x03
TILT_UP, TILT_DOWN, TILT_STOP = 0x01, 0x02, 0x03

MAX_PAN_SPEED = 24
MAX_TILT_SPEED = 20
MAX_ZOOM_SPEED = 7


def _clamp(value: int, lo: int, hi: int) -> int:
    return max(lo, min(hi, int(round(value))))


def pan_tilt(pan_dir: int, tilt_dir: int, pan_speed: int, tilt_speed: int) -> bytes:
    """81 01 06 01 [pan_speed] [tilt_speed] [pan_dir] [tilt_dir] FF"""
    return bytes([0x81, 0x01, 0x06, 0x01,
                  _clamp(pan_speed, 1, MAX_PAN_SPEED),
                  _clamp(tilt_speed, 1, MAX_TILT_SPEED),
                  pan_dir, tilt_dir, 0xFF])


def pan_tilt_stop() -> bytes:
    return bytes([0x81, 0x01, 0x06, 0x01, 0x01, 0x01, PAN_STOP, TILT_STOP, 0xFF])


def zoom(direction: str, speed: int) -> bytes:
    """direction 'in' (tele, 0x2s) / 'out' (wide, 0x3s) / anything else = stop."""
    s = _clamp(speed, 1, MAX_ZOOM_SPEED)
    if direction == "in":
        byte = 0x20 | s
    elif direction == "out":
        byte = 0x30 | s
    else:
        byte = 0x00
    return bytes([0x81, 0x01, 0x04, 0x07, byte, 0xFF])


def zoom_stop() -> bytes:
    return bytes([0x81, 0x01, 0x04, 0x07, 0x00, 0xFF])


def home() -> bytes:
    return bytes([0x81, 0x01, 0x06, 0x04, 0xFF])


class ViscaSender:
    """One shared non-blocking UDP socket; sendto per camera IP."""

    def __init__(self, port: int = DEFAULT_PORT) -> None:
        self.port = port
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._sock.setblocking(False)

    def send(self, ip: str, command: bytes) -> None:
        try:
            self._sock.sendto(command, (ip, self.port))
        except OSError as exc:
            log.debug("VISCA send to %s failed: %s", ip, exc)

    def close(self) -> None:
        self._sock.close()
